File: BusquedaLineal.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int busquedaLineal(int arr[], int n, int k) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == k) {
            return i;
        }
    }
    return -1;
}

int main() {
    FILE *archivo;
    int *numeros;
    int n, k, pos;

    archivo = fopen("datos.txt", "r");
    if (archivo == NULL) {
        printf("Error al abrir el archivo.\n");
        return 1;
    }

    printf("Ingrese la cantidad de datos a leer (n): ");
    scanf("%d", &n);

    printf("Ingrese el numero a buscar (k): ");
    scanf("%d", &k);

    numeros = (int*) malloc(n * sizeof(int));
    if (numeros == NULL) {
        printf("Error al reservar memoria.\n");
        fclose(archivo);
        return 1;
    }

    for (int i = 0; i < n; i++) {
        if (fscanf(archivo, "%d", &numeros[i]) != 1) {
            printf("No hay suficientes datos en el archivo.\n");
            free(numeros);
            fclose(archivo);
            return 1;
        }
    }
    fclose(archivo);

    printf("\n");

    clock_t inicio = clock();
    pos = busquedaLineal(numeros, n, k);
    clock_t fin = clock();

    double tiempo = (double)(fin - inicio) / CLOCKS_PER_SEC;

    if (pos != -1) {
        printf("El numero %d se encontro en la posicion %d.\n", k, pos);
    } else {
        printf("El numero %d NO se encontro.\n", k);
    }

    printf("Tiempo de ejecucion de la busqueda lineal: %f segundos\n", tiempo);

    free(numeros);
    return 0;
}

File: BusquedaLineal.py
import time


def busqueda_lineal(arr, k):
    for indice, valor in enumerate(arr):
        if valor == k:
            return indice
    return -1


def main():
    try:

        with open('datos.txt', 'r') as archivo:

            todos_los_numeros = [int(num) for num in archivo.read().split()]

        n = int(input("Ingrese la cantidad de datos a leer (n): "))
        k = int(input("Ingrese el numero a buscar (k): "))

        if n > len(todos_los_numeros):
            print("No hay suficientes datos en el archivo.")
            return

        numeros = todos_los_numeros[:n]

        print("\nIniciando búsqueda...")

        inicio = time.perf_counter()
        pos = busqueda_lineal(numeros, k)
        fin = time.perf_counter()

        tiempo = fin - inicio

        if pos != -1:

            print(f"El numero {k} se encontro en la posicion {pos}.")
        else:
            print(f"El numero {k} NO se encontro.")

        print(f"Tiempo de ejecucion de la busqueda lineal: {tiempo:.6f} segundos")

    except FileNotFoundError:
        print("Error: El archivo 'datos.txt' no se encontró en el directorio.")
    except ValueError:
        print("Error: Por favor, ingrese un número entero válido.")
    except Exception as e:
        print(f"Ocurrió un error inesperado: {e}")


if __name__ == "__main__":
    main()
